File: web/app/data/geo.ts
// FlyBy geo — great-circle distance + bbox helpers.
//
// Ported from src/geo.cpp so the web app and the firmware
// agree on what "nearest" means.

const EARTH_RADIUS_KM = 6371.0088 // WGS-84 mean radius
const KM_PER_DEG_LAT = 111.32
const DEG_TO_RAD = Math.PI / 180

export interface BBox {
  lamin: number
  lomin: number
  lamax: number
  lomax: number
}

// Great-circle distance between two (lat, lon) points in kilometers.
export function haversineKm(lat1: number, lon1: number, lat2: number, lon2: number): number {
  let dLat = (lat2 - lat1) * DEG_TO_RAD
  let dLon = (lon2 - lon1) * DEG_TO_RAD
  let s1 = Math.sin(dLat / 2)
  let s2 = Math.sin(dLon / 2)
  let a =
    s1 * s1 +
    Math.cos(lat1 * DEG_TO_RAD) * Math.cos(lat2 * DEG_TO_RAD) * s2 * s2
  return 2 * EARTH_RADIUS_KM * Math.asin(Math.sqrt(a))
}

// Approximate lat/lon bounding box around (lat, lon) covering the given
// radius. The box is over-inclusive (a square in lat/lon space, not a
// true circle), which is what OpenSky's bbox query expects.
//
// Latitude is clamped to [-90, 90] and longitude to [-180, 180] at the
// edges — at the poles or near the antimeridian, a naïve bbox would
// spill out of valid coordinate space and OpenSky would reject it.
// Clamping gives a truncated (but valid) box; good enough for a display
// that realistically isn't used from inside the Arctic Circle.
export function bboxFor(lat: number, lon: number, radiusKm: number): BBox {
  let dLat = radiusKm / KM_PER_DEG_LAT
  let cosLat = Math.cos(lat * DEG_TO_RAD)
  let dLon = cosLat > 1e-9 ? radiusKm / (KM_PER_DEG_LAT * cosLat) : 180
  return {
    lamin: Math.max(-90, lat - dLat),
    lomin: Math.max(-180, lon - dLon),
    lamax: Math.min(90, lat + dLat),
    lomax: Math.min(180, lon + dLon),
  }
}

File: src/geo.cpp
#include "geo.h"

#include <cmath>

namespace geo {
namespace {

constexpr double kEarthRadiusKm = 6371.0088;  // WGS-84 mean radius
constexpr double kPi = 3.14159265358979323846;
constexpr double kDegToRad = kPi / 180.0;

// 1 degree of latitude ≈ this many km (constant globally).
constexpr double kKmPerDegLat = 111.32;

}  // namespace

double haversine_km(double lat1, double lon1, double lat2, double lon2) {
  const double dLat = (lat2 - lat1) * kDegToRad;
  const double dLon = (lon2 - lon1) * kDegToRad;
  const double s1 = std::sin(dLat / 2.0);
  const double s2 = std::sin(dLon / 2.0);
  const double a = s1 * s1 +
                   std::cos(lat1 * kDegToRad) * std::cos(lat2 * kDegToRad) *
                       s2 * s2;
  return 2.0 * kEarthRadiusKm * std::asin(std::sqrt(a));
}

BBox bbox_for(double lat, double lon, double radius_km) {
  // Latitude is uniform; longitude shrinks toward the poles.
  const double dLat = radius_km / kKmPerDegLat;
  const double cosLat = std::cos(lat * kDegToRad);
  // Guard against division-by-zero at the poles. cos(90°) is ~6e-17, not 0,
  // but we still want a sensible bbox if someone tests with lat = 90.
  const double dLon = (cosLat > 1e-9)
                          ? (radius_km / (kKmPerDegLat * cosLat))
                          : 180.0;
  return {lat - dLat, lon - dLon, lat + dLat, lon + dLon};
}

double bearing_deg(double lat1, double lon1, double lat2, double lon2) {
  const double phi1 = lat1 * kDegToRad;
  const double phi2 = lat2 * kDegToRad;
  const double dLon = (lon2 - lon1) * kDegToRad;
  const double y = std::sin(dLon) * std::cos(phi2);
  const double x = std::cos(phi1) * std::sin(phi2) -
                   std::sin(phi1) * std::cos(phi2) * std::cos(dLon);
  double deg = std::atan2(y, x) / kDegToRad;
  if (deg < 0) deg += 360.0;
  return deg;
}

const char* bearing_to_compass(double bearing) {
  // 8 sectors, 45° each, offset by 22.5° so N is 0±22.5°.
  while (bearing < 0) bearing += 360.0;
  while (bearing >= 360.0) bearing -= 360.0;
  int sector = static_cast<int>((bearing + 22.5) / 45.0) % 8;
  static const char* labels[] = {"N", "NO", "O", "SO", "S", "SW", "W", "NW"};
  return labels[sector];
}

}  // namespace geo
